File: Partition Labels/Leetcode_763.py
# O(N) time, O(1) space
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        last_occurence = {char: i for i, char in enumerate(S)}
        ans, cur_start, cur_break = [], 0, last_occurence[S[0]]
        for idx, char in enumerate(S):
            cur_break = max(cur_break, last_occurence[char])
            if cur_break == idx:
                ans += [cur_break-cur_start+1]
                cur_start = idx+1
        return ans
    

# O(N) time, N=len(S)
# O(L) space, L = length of alphabet
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        d = {}
        for i, char in enumerate(S):
            d[char] = i
        ans = []
        start = 0
        cur, finished = set(), set()
        for i, char in enumerate(S):
            if d[char] == i: 
                finished.add(char)
            cur.add(char)
            if cur == finished:
                ans += [i+1-start]
                start = i+1
        return ans

File: Partition Labels/Leetcode_763.rs
use std::collections::HashMap;

impl Solution {
    pub fn partition_labels(s: String) -> Vec<i32> {
        let n: usize = s.len();
        
        // O(L) space, L=size(alphabet)
        let S: Vec<char> =(&s).chars().collect();
        let mut d: HashMap<char, usize> = HashMap::new();
        for i in 0..n{
            d.insert(S[i], i);
        }
        let mut res: Vec<i32> = Vec::new();
        let mut cur_start: usize = 0;
        let mut cur_end: usize = d[&S[0]];
        
        // O(N) time
        for i in 0..n{
            if d[&S[i]] > cur_end {cur_end = d[&S[i]];}
            if i==cur_end {
                res.push((cur_end-cur_start+1) as i32);
                cur_start = i+1;
            }
        }; 
        return res;
        } 
    }
